metagraph/src/annotation/binary_matrix/row_sparse_disk/row_sparse_disk.cpp

```diff
@@ -0,0 +1,141 @@
+#include "row_sparse_disk.hpp"
+
+#include "common/logger.hpp"
+#include "common/utils/file_utils.hpp"
+#include "common/ifstream_with_name_and_offset.hpp"
+#include "graph/representation/succinct/boss.hpp"
+
+
+namespace mtg {
+namespace annot {
+namespace binmat {
+
+using mtg::common::logger;
+
+bool RowSparseDisk::Impl::get(Row row, Column column) const {
+    SetBitPositions set_bits = get_row(row);
+    SetBitPositions::iterator v = std::lower_bound(set_bits.begin(), set_bits.end(), column);
+    return v != set_bits.end() && *v == column;
+}
+
+std::vector<BinaryMatrix::Row> RowSparseDisk::Impl::get_column(uint64_t num_rows,
+                                                               Column column) const {
+    std::vector<Row> result;
+    for (Row row = 0; row < num_rows; ++row) {
+        if (get(row, column))
+            result.push_back(row);
+    }
+    return result;
+}
+
+BinaryMatrix::SetBitPositions RowSparseDisk::Impl::get_row(Row row) const {
+    assert(boundary_[boundary_.size() - 1] == 1);
+    uint64_t start_idx = row == 0 ? 0 : boundary_.select1(row) + 1;
+    uint64_t end_idx = boundary_.next1(start_idx);
+    SetBitPositions result;
+    result.reserve(end_idx - start_idx);
+    // In each row, the first value in `set_bits_` stores the first set bit,
+    // and all next ones store deltas pos[i] - pos[i-1].
+    uint64_t last = 0;
+    for (uint64_t i = start_idx; i != end_idx; ++i) {
+        result.push_back(last += set_bits_[i - row]);
+    }
+    return result;
+}
+
+std::vector<BinaryMatrix::SetBitPositions>
+RowSparseDisk::Impl::get_rows(const std::vector<Row> &row_ids) const {
+    std::vector<SetBitPositions> rows(row_ids.size());
+
+    for (size_t i = 0; i < row_ids.size(); ++i) {
+        rows[i] = get_row(row_ids[i]);
+    }
+
+    return rows;
+}
+
+bool RowSparseDisk::load(std::istream &f) {
+    auto _f = dynamic_cast<mtg::common::IfstreamWithNameAndOffset *>(&f);
+    assert(_f);
+    try {
+        num_columns_ = load_number(f);
+
+        auto boundary_start = load_number(f);
+
+        _f->set_offset(static_cast<uint64_t>(f.tellg()));
+
+        int_vector_buffer_params.filename = _f->get_name();
+        int_vector_buffer_params.offset = _f->get_offset();
+
+        f.seekg(-sizeof(std::streampos), ios_base::end);
+
+        f.seekg(boundary_start, ios_base::beg);
+        boundary_.load(f);
+
+        num_rows_ = boundary_.num_set_bits();
+
+    } catch (...) {
+        return false;
+    }
+
+    return true;
+}
+
+void RowSparseDisk::serialize(std::ostream & /*f*/) const {
+    throw std::runtime_error(
+            "RowSparseDisk::serialize not supproted, use static variant instead");
+}
```

Comment:
Well, this is a good question and in fact at the beginning of my work I wanted to do this like other annotators, i.e. building in the constructor and then using serialize.
Maybe it would be possible but as it is disk-based annotation it would require that constructor write to disk. I mean the constructor would do something similar to what currently is done in static serialize, i.e. it would write int_vector_buffer with rows to the disk.
Then non-static serialize would rewrite it from disk to full annotation. In my opinion, not a good choice since in practice it would mean writing row content on disk twice.
I guess the alternative would be to store rows in memory when this loading ctor is used, but it doesn't seem to make sence to have a ctor that loads data to memory for a class that is disk-based.
Do you see another, better, way to do it more consistently or are we leaving it as is?
Maybe there are also some other, more implementation-related, issues to be considered here, but I think the one mentioned is quite important.

Hmm, now I see that you probably meant having both static and non-static variants. I guess that would be possible. I will try to add this.

Reply:
I implemented a non-static variant. The int_vector_buffer part of the annotation is just binary rewritten (that way I don't need to read the data with one int_vector_buffer and store it with another one, which would require to have a file name, which is not available in the non-static serialize and would require similar trick like in load (i.e. IfstreamWithName). Let me know if that's fine.